main.go

```diff
@@ -401,49 +401,175 @@ func connectAction(ctx *cli.Context) error {
 	return nil
 }
 
-// disconnectAction tries to stop rhscd service, disconnect from Red Hat Insights and finally
+// setupFormatOption ensures the user has supplied a correct `--format` flag
+// and set values in uiSettings, when JSON format is used.
+func setupFormatOption(ctx *cli.Context) error {
+	// This is run after the `app.Before()` has been run,
+	// the uiSettings is already set up for us to modify.
+	format := ctx.String("format")
+	switch format {
+	case "":
+		return nil
+	case "json":
+		uiSettings.isMachineReadable = true
+		uiSettings.isRich = false
+		return nil
+	default:
+		err := fmt.Errorf(
+			"unsupported format: %s (supported formats: %s)",
+			format,
+			`"json"`,
+		)
+		return cli.Exit(err, 1)
+	}
+}
+
+// DisconnectResult is structure holding information about result of
+// disconnect command. The result could be printed in machine-readable format.
+type DisconnectResult struct {
+	Hostname                  string `json:"hostname"`
+	HostnameError             string `json:"hostname_error,omitempty"`
+	UID                       int    `json:"uid"`
+	UIDError                  string `json:"uid_error,omitempty"`
+	RHSMDisconnected          bool   `json:"rhsm_disconnected"`
+	RHSMDisconnectedError     string `json:"rhsm_disconnect_error,omitempty"`
+	InsightsDisconnected      bool   `json:"insights_disconnected"`
+	InsightsDisconnectedError string `json:"insights_disconnected_error,omitempty"`
+	YggdrasilStopped          bool   `json:"yggdrasil_stopped"`
+	YggdrasilStoppedError     string `json:"yggdrasil_stopped_error,omitempty"`
+	exitCode                  int
+}
+
+// printJSONDisconnectResult tries to print the result of disconnect as JSON to stdout.
+// When marshaling of systemStatus fails, then error is returned
+func printJSONDisconnectResult(disconnectResult *DisconnectResult) error {
+	data, err := json.MarshalIndent(disconnectResult, "", "    ")
+	if err != nil {
+		return err
+	}
+	fmt.Println(string(data))
+	return nil
+}
+
+// beforeDisconnectAction ensures the used has supplied a correct `--format` flag
+func beforeDisconnectAction(ctx *cli.Context) error {
+	return setupFormatOption(ctx)
+}
+
+// disconnectAction tries to stop rhscd service, disconnect from Red Hat Insights, and finally
 // it unregister system from Red Hat Subscription Management
 func disconnectAction(ctx *cli.Context) error {
+	var disconnectResult DisconnectResult
+	var machineReadablePrintFunc func(disconnectResult *DisconnectResult) error
+
+	format := ctx.String("format")
+	switch format {
+	case "json":
+		machineReadablePrintFunc = printJSONDisconnectResult
+	default:
+		break
+	}
+
+	// When printing of status is requested, then print machine-readable file format
+	// at the end of this function
+	if uiSettings.isMachineReadable {
+		defer func(disconnectResult *DisconnectResult) {
```

Comment:
You cannot always declare `defer` function, because when machine readable output is not used, we print some progress messages, spinners, etc. We cannot print everything at the end.